main.tf

```diff
@@ -4,6 +4,7 @@ terraform {
 
 locals {
   max_subnet_length = "${max(length(var.private_subnets), length(var.elasticache_subnets), length(var.database_subnets), length(var.redshift_subnets))}"
+  nat_gateway_count = "${var.single_nat_gateway ? 1 : local.max_subnet_length}"
```

Comment:
`max_subnet_length` gets longest of all 4 subnet types, while NAT should only be applicable to `private_subnets`

Reply:
From what I can see, that's not how it works currently. All the private-type subnets, including the "extras" (database, redshift, elasticache), are associated to the "private" route tables. This means they all route through the NAT Gateway. The number of NAT Gateways therefore needs to account for all the private-type subnets, which is what `max_subnet_length` does.